mqtt/mqtt-bridge/src/persist/waking_state/waking_map.rs

```diff
@@ -0,0 +1,67 @@
+use std::{cmp::min, collections::HashMap, collections::VecDeque, task::Waker};
+
+use mqtt3::proto::Publication;
+use tracing::error;
+
+use crate::persist::{waking_state::StreamWakeableState, Key, PersistError};
+
+/// Responsible for waking waiting streams when new elements are added.
+///
+/// Exposes a get method for retrieving a count of elements in order of insertion.
+/// When elements are retrieved they are moved to the in flight collection.
+pub struct WakingMap {
+    queue: VecDeque<(Key, Publication)>,
+    in_flight: HashMap<Key, Publication>,
+    waker: Option<Waker>,
+}
+
+impl WakingMap {
+    pub fn new() -> Self {
+        let queue: VecDeque<(Key, Publication)> = VecDeque::new();
+        let in_flight = HashMap::new();
+
+        WakingMap {
+            queue,
+            in_flight,
+            waker: None,
+        }
+    }
+}
+
+impl StreamWakeableState for WakingMap {
+    fn insert(&mut self, key: Key, value: Publication) -> Result<(), PersistError> {
+        self.queue.push_back((key, value));
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+
+        Ok(())
+    }
+
+    fn get(&mut self, count: usize) -> Vec<(Key, Publication)> {
```

Comment:
get usually means get *one* item from the collection. `fn batch()` sounds better